### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by ThmsMoritz**
Added methods incremental_input_before and incremental_input_in_steps to allow for more dynamic benchmarks execution.

Also added a method showcasing a non-functional case of adding 21 elementAs at once.

### CODE REVIEW

**Review by gschennersie — DISMISSED** [review]

**Comment by ThmsMoritz** [comment]
Currently 49 single, 27 double and 16 triple benchmarks are run (meaning the iterations/how often objects were added).

Most cases will still result in a timeout, which hopefully will change after the rackDouble bug is fixed. The automated benchmarks may require a small bit of adjusting after that to better cover then still problematic cases.


Additionally to the automated benchmark method itself (run_incremental_benchmarks), following further methods were added:

- run_incremental_benchmark_instance: running the respective benchmarks instances, called from run_incremental_benchmarks

- add_elements: adding new elements to a specified configuration

- extend_incrementall_wrapper: a wrapper for the extend_incrementally method making it easier to call the method on a new thread.

- setup_bm_instance: creating a new configuration with specified numbers of different elements

File: asp_interactive_configuration/benchmarks/benchmark.py [hunk 1]
```diff
@@ -100,6 +101,36 @@ def new_iconf():
     racks_kb = OOASPKnowledgeBase.from_file("racks_v1",settings.racks_example_kb)
     return InteractiveConfigurator(racks_kb,"i1",[settings.racks_example_constraints])
 
+def setup_bm_instance(elemAs, elemBs, elemCs, elemDs):
```
Comment (rtaupe) [inline asp_interactive_configuration/benchmarks/benchmark.py:104]: Please provide docstrings for the newly introduced methods

File: asp_interactive_configuration/benchmarks/benchmark.py [hunk 2]
```diff
@@ -132,6 +163,159 @@ def options_object(ne):
...
+    if thread.is_alive():
+        threading.Event().set()
+        print("The extend_incrementally method timed out after", run_in_seconds, "seconds. \n")
+        return
+
+    iconf.select_found_configuration()
+    print(str(iconf._time_grounding + iconf._time_solving) + "\n")
+    #found[0].save_png("benchmarks/results", "-incremental" + str(20))
+
+def run_incremental_benchmarks(run_in_seconds: int):
```
Comment (rtaupe) [inline asp_interactive_configuration/benchmarks/benchmark.py:313]: This (and other) method(s) is never called. If this is the main method running all benchmarks, please add a sample call at the end of the file. In the end, there should be no methods that are not called from anywhere.

File: asp_interactive_configuration/benchmarks/benchmark.py [hunk 3]
```diff
@@ -132,6 +163,159 @@ def options_object(ne):
...
+
+        found.save_png()
+        iteration_nr += 1
+    return iconf
+
+def run_incremental_benchmark_instance(iconf:InteractiveConfigurator, run_in_seconds):
+
+    #found = []
+    #if interrupted; console wont work anymore
+    #thread = threading.Thread(target=extend_incrementally_wrapper, args=(iconf, found))
```
Comment (rtaupe) [inline asp_interactive_configuration/benchmarks/benchmark.py]: Please do not include commented-out code.

File: asp_interactive_configuration/benchmarks/benchmark.py [hunk 4]
```diff
@@ -132,6 +163,159 @@ def options_object(ne):
...
+        print(iconf._time_grounding + iconf._time_solving)
+
+        found.save_png()
+        iteration_nr += 1
+    return iconf
+
+def run_incremental_benchmark_instance(iconf:InteractiveConfigurator, run_in_seconds):
+
+    #found = []
+    #if interrupted; console wont work anymore
```
Comment (rtaupe) [inline asp_interactive_configuration/benchmarks/benchmark.py]: Please have a look at https://github.com/potassco/clingo/issues/151 to improve handling of interrupts and/or timeouts.

**Review by rtaupe — CHANGES_REQUESTED** [review]
Thank you for these contributions!
Please address the comments and also update the description of the pull request (it seems that much more than just adding the methods `incremental_input_before` and `incremental_input_in_steps` has happened).

**Comment by ThmsMoritz** [comment]
Addressed mentioned issues, still missing the fix for proper handling of the timeout.

**Comment by ThmsMoritz** [comment]
Still open:

- Proper termination of solve thread in `interactive._next_solution()` (currently iterator is implicitly waiting for termination of solving)
- Dynamic parametrizable timeout, hand over time out limit (reduced from iteration to iteration) from `interactive.extend_incrementally()`
- Prevent further iterations in case of timeout

**Comment by rtaupe** [comment]
Closed, will be superseded by a declarative definition of benchmarks.

**Pull request closed without merging**